### PR TITLE
ValueMapperTransformer: Added support for loading map from file through dataview.

### LINKED ISSUE #2162: Add support for vector as input/output in ValueMappingTransformer
Currently, ValueMappingTransformer https://github.com/dotnet/machinelearning/blob/ef638f480a2be117b181ea52a1786e391244e5d2/src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs#L256 does not take vector as input. It works on only on scalar input.

To enable Text/NLP scenario (#747) in TensorFlowTransform where model's expected input is vector of integers (indexes into dictionary), following needs to be done. Note that this is only needed if model accepts vector of integers as input.

1. Tokenize the string into words.
2. Map the words to indexes into the dictionary provided with the TF model.
3. If the model is Seq2Seq model, take the output from the model which is again vector of integers and convert it back to vector of text.

No. 1 can be done using ML.Net's WordTokenizer.
No. 2 and 3 can be done using ML.Net's ValueMappingTransformer. However, it does not currently supports vector as input/output. Furthermore, the Estimator https://github.com/dotnet/machinelearning/blob/ef638f480a2be117b181ea52a1786e391244e5d2/src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs#L42 of ValueMappingTransformer also does not expose interface to load dictionary from file.

### PR DESCRIPTION
**Pull request by zeahmed**
This fixes #2162. This is the 2nd part where the support for loading the mapping from file is added.

### CODE REVIEW

**Comment by singlis** [comment]
>     public sealed class ValueMappingEstimator<TKey, TValue> : TrivialEstimator<ValueMappingTransformer<TKey, TValue>>

Can ValueMappingEstimator<TKey, TValue> derive from ValueMappingEstimator? You could then remove the duplicate GetOutputSchema.

---
Refers to: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs:91 in 47b757a. [](commit_id = 47b757a9f93ee2fac74e06f4a9efc25627d1bd72, deletion_comment = False)

**Comment by singlis** [comment]
>     public sealed class ValueMappingEstimator<TKey, TValue> : TrivialEstimator<ValueMappingTransformer<TKey, TValue>>

So I think you can remove ValueMappingTransformer<TKey, TValue> and have ValueMappingEstimator<TKey, TValue> derive from ValueMappingEstimator which derives from TrivialEstimator<ValueMappingTransform>. ValueMappingTransform only takes in an IDataView. ValueMappingEstimator<TKey, TValue> will have to create a dataview from the passed in keys and values, which should be a matter of calling the CreateDataVIew static function. Its a bit more involved of a change, but it would reduce the classes.

What do you think?

---
In reply to: [457672521](https://github.com/dotnet/machinelearning/pull/2232#issuecomment-457672521) [](ancestors = 457672521,457435302)

---
Refers to: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs:91 in 47b757a. [](commit_id = 47b757a9f93ee2fac74e06f4a9efc25627d1bd72, deletion_comment = False)

**Comment by zeahmed** [comment]
>     public sealed class ValueMappingEstimator<TKey, TValue> : TrivialEstimator<ValueMappingTransformer<TKey, TValue>>

That pretty much I was thinking to do.

---
In reply to: [457672521](https://github.com/dotnet/machinelearning/pull/2232#issuecomment-457672521) [](ancestors = 457672521,457435302)

---
Refers to: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs:91 in 47b757a. [](commit_id = 47b757a9f93ee2fac74e06f4a9efc25627d1bd72, deletion_comment = False)

File: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs [hunk 1]
```diff
@@ -32,14 +32,63 @@
 
 namespace Microsoft.ML.Transforms.Conversions
 {
+    /// <summary>
+    /// The ValueMappingEstimator is a 1-1 mapping from a key to value. This particular class load the mappings from an <see cref="IDataView"/>.
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs]: >key [](start = 58, length = 3)

reference to KeyType

File: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs [hunk 2]
```diff
@@ -32,14 +32,63 @@
...
+        private readonly (string input, string output)[] _columns;
+
+        /// <summary>
+        /// Constructs the ValueMappingEstimator, key type -> value type mapping
+        /// </summary>
+        /// <param name="env">The environment to use.</param>
+        /// <param name="lookupMap">An instance of <see cref="IDataView"/> that contains the key and value columns.</param>
+        /// <param name="keyColumn">Name of the key column in <paramref name="lookupMap"/>.</param>
+        /// <param name="valueColumn">Name of the value column in <paramref name="lookupMap"/>.</param>
+        /// <param name="columns">The list of columns to apply.</param>
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs]: >The list of columns to apply [](start = 34, length = 28)

The list of names of the input columns to apply the transformation, and the name of the resulting column.

File: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs [hunk 3]
```diff
@@ -32,14 +32,63 @@
...
+        public ValueMappingEstimator(IHostEnvironment env, IDataView lookupMap, string keyColumn, string valueColumn, params (string input, string output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingEstimator)),
+                    new ValueMappingTransformer(env, lookupMap, keyColumn, valueColumn, columns))
+        {
+            _columns = columns;
+        }
+
+        public override SchemaShape GetOutputSchema(SchemaShape inputSchema)
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs]: >public [](start = 8, length = 6)

xml doc

File: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs [hunk 4]
```diff
@@ -32,14 +32,63 @@
 
 namespace Microsoft.ML.Transforms.Conversions
 {
+    /// <summary>
+    /// The ValueMappingEstimator is a 1-1 mapping from a key to value. This particular class load the mappings from an <see cref="IDataView"/>.
+    /// This gives user the flexibility to load the mapping from file instead of using IEnumerable in <see cref="ValueMappingEstimator{TKey, TValue}"/>
+    /// </summary>
+    public class ValueMappingEstimator : TrivialEstimator<ValueMappingTransformer>
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs]: rename the file to ValueMapping if both the estimator and the transformer are on the same file.

**Comment by sfilipi** [comment]
>     public sealed class ValueMappingEstimator<TKey, TValue> : TrivialEstimator<ValueMappingTransformer<TKey, TValue>>

+1 :)

---
In reply to: [457693072](https://github.com/dotnet/machinelearning/pull/2232#issuecomment-457693072) [](ancestors = 457693072,457672521,457435302)

---
Refers to: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs:91 in 47b757a. [](commit_id = 47b757a9f93ee2fac74e06f4a9efc25627d1bd72, deletion_comment = False)

File: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs (hunk 1 above)
Comment (zeahmed) [reply src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs]: Here term `Key` is used as general term as used in dictionary/map. Its not referencing ML.Net KeyType specifically here. Let me know if you meant something else?

---
In reply to: [251594412](https://github.com/dotnet/machinelearning/pull/2232#discussion_r251594412) [](ancestors = 251594412)

File: src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs (hunk 4 above)
Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs]: Did this get resolved? The file is still ValueMappingTransformer.

---
In reply to: [251596169](https://github.com/dotnet/machinelearning/pull/2232#discussion_r251596169) [](ancestors = 251596169)

Comment (zeahmed) [reply src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs]: Resolved if you see the sixth commit.

---
In reply to: [251634144](https://github.com/dotnet/machinelearning/pull/2232#discussion_r251634144) [](ancestors = 251634144,251596169)

**Comment by zeahmed** [comment]
Thanks @singlis and @sfilipi  for the useful comments!

**Pull request merged**